runtime/src/accounts_index.rs

```diff
@@ -581,6 +881,38 @@ impl<T: 'static + Clone> AccountsIndex<T> {
         max_root
     }
 
+    pub fn update_secondary_indexes(
+        &self,
+        pubkey: &Pubkey,
+        slot: Slot,
+        account_owner: &Pubkey,
+        account_data: &[u8],
+        account_indexes: &HashSet<AccountIndex>,
+    ) {
+        if account_indexes.contains(&AccountIndex::ProgramId) {
+            self.program_id_index.insert(account_owner, pubkey, slot);
+        }
```

Comment:
maybe add comment like this?

``` rust
// we skip updating even if given pubkey's account is reset to Account::Default due to `lamports == 0`.
// Skipping means not updating secondary index to mark the account as missing.
// This doesn't introduce false positives (= wrong index hit) because reset accounts has already
// previous updates at least, otherwise it would be purged altogether via purge_keys_exact in `fn clean_accounts`,
// including the secondary indexes.
```

Reply:
@ryoqun I think the false positives warning is helpful, but the justification needs a little tweaking. This is my thoughts:

```
// Note because of the below check below on the account data length, when an
        // account hits zero lamports and is reset to Account::Default, then we skip
        // the below updates to the secondary indexes.
        //
        // Skipping means not updating secondary index to mark the account as missing.
        // This doesn't introduce false positives during a scan because the caller to scan
        // provides the ancestors to check. So even if a zero-lamport account is not yet
        // removed from the secondary index, the scan function will:
        // 1) consult the primary index via `get(&pubkey, Some(ancestors), max_root)`
        // and find the zero-lamport version
        // 2) When the fetch from storage occurs, it will return nothing because the account
        // is default in that storage entry
```

Reply:
yeah, perfect rewrite. :) Thanks for polishing my jottings. :)